### ISSUE TITLE
C++ Style Guide: Incorrect information about rvalue references

### ISSUE THREAD

**Issue by cmazakas** [google/styleguide#355]
> Rvalue references are a type of reference that can only bind to temporary objects. 

This is only partially correct.

Temporaries can bind to either `value const&` or `value_type&&`. The difference being, rvalue references are mutable and can bind to temporaries.

Rvalue references also bind entirely to rvalues.

Consider,
```
std::string str = "kitty cat";
std::string&& s = std::move(str);
```

Here you can see, `str` is not a temporary but rather, `std::move(str)` is an xvalue expression which has identity.

The wording should be changed to something such as,

Rvalue references are mutable references that represent resource reuse, i.e. the contents of the class are expected to be safely pilfered by consuming.

It's important to note that both xvalue and prvalue expressions are rvalue'd.

**Comment by as** [comment]
https://github.com/google/styleguide/blob/1b206ee36263b16ec18f7b2f86a2b770b7490844/cppguide.html#L2396

**Comment by tituswinters** [comment]
That's all true, but the style guide is pretty expressly written in a non-language-lawyer fashion, and decision sections try to be short and conceptually clear (rather than hit all of the language details).  If there's a very small wording change, that might be fine.  Otherwise, we're likely to leave it as-is. 

(There's also some internal work to revise the approach on rvalue refs right now, so this may get swept up in that change anyway.)

**Comment by cmazakas** [comment]
Heh, that's fair.

For the revised guide, it might be better to break things down in terms of reference binding. That's what a lot of these value categories seem to be centered around.

For example, I was unaware that temporaries even bound to const references until someone mentioned it on the cpplang slack team.

So instead of an rvalue references section, it'd be more like a reference binding section. We could then cover a lot of topics/useful edge cases in one place.

For example, why we `std::move` even `value_type&& val`, what happens if we just use `val` as-is.  If possible, I'd love to contribute documentation on C++ in this regard.

cppreference has pretty good breakdowns on the value categories of expressions now as well. For example, glvalue'd expressions = identity

Breaking it up into a Q&A kind of format might work as well. For example, "why is T().x() xvalue'd?"

**Comment by tituswinters** [comment]
That all sounds like good content, but more "How does the language work" than "How should you restrict yourself/structure your code when writing?" We don't have a lot of good tutorial material (that's a huge task), and the style guide is definitely not the place for it.

But if there's good / accessible summary of value category that isn't too lawyer-y, maybe the right answer is just to link to that from this section?

**Comment by cmazakas** [comment]
For one, I would love to write accessible and in-depth documentation about C++ and its more intricate nooks and crannies. These things, while pedantic, ultimately ease the burden of reading and writing C++ and empower the programmer.

Proposed wording change:
> Rvalue references are a reference type that represent resource reuse. Unlike lvalue references which represent ownership, rvalue references broadcast that the resources of the class/type _may_ be reused by consuming code (and as such are mutable). The syntax is similar to traditional reference syntax. For example, void f(string&& s); declares a function whose argument is an rvalue reference to a string.
> Rvalue references may bind to temporaries but also to any `std::move`'d class that is not `const`-qualified and supports move construction/assignment.

> *Cons:*
>  * Rvalue references are a relatively new feature (introduced as part of C++11), and not yet widely understood. Rules like reference collapsing, and automatic synthesis of move constructors, are complicated.
> * `std::move` is ultimately "just a cast" and can make generic code confusing for the reader
> * Rvalue references require an in-depth knowledge of the value categories of C++ and can be easily misused or mislead inexperienced readers
